### DISCUSSION TITLE
Add Tailwind-esque Style Parsing

### DISCUSSION THREAD

**Discussion by lewxdev** [tamagui/tamagui#1577] [General]
Historically, I haven't been the biggest fan of tailwind. But working with tamagui, I'm starting to see some of the merit to it: being able to separate out style props from component props could help improve the DX in specific use cases.

## What I'm proposing

Instead of a component like...
```tsx
<Input
  autoCorrect={false}
  backgroundColor="$slate2"
  borderColor="$slate6"
  borderWidth={3}
  color="$slate9"
  placeholder="Enter your name"
  size="$5"
  value={firstName}
  onChangeText={setFirstName}
/>
```

We could do...
```tsx
<Input
  style="bg-slate2 boc-slate6 bw-3 col-slate9 size-5"
  autoCorrect={false}
  placeholder="Enter your name"
  value={firstName}
  onChangeText={setFirstName}
/>
```

- Uses the [recommended/default shorthands](https://github.com/tamagui/tamagui/blob/master/packages/shorthands/src/index.tsx), as prefixes
- Could potentially allow for integrating with the tailwind ecosystem of tools and plugins (and/or copy their homework more easily, see #1573)
- Reduces unnecessary verbosity when it comes to styling

At first this can be an experimental opt-in feature. The `style` prop on `TamaguiComponent` could be a union of `StyleProp | string` when the feature is enabled which parses and handles styling just as the props would

**Comment by cerpinn** [comment]
Related: https://github.com/tamagui/tamagui/pull/234

**Comment by natew** [comment]
Converting to discussion. I agree that separate style from functional props can be nice, but I dislike string based. While you list one upside, there are about 10 downsides.

And you can get the same upside by throwing styles into an object, for example.

We have a nice VSCode plugin btw for showing colors and token values.

And also you don't want prop sorting in Tamagui, as order is significant, and actually solves one of Tailwinds biggest problems which is the lack of control over merging (yea tw merge solves this partially, but now you're still back to dealing with strings and overriding things granularly is a total pain, whereas Tamagui you have variants which expand, and then you can override granularly afterwards easily).

All that said, if you want a style prop like `sx` you can do it pretty easily with Tamagui, the only tricky part would be the types. You'd need to make your own base view, then you can make a variant with `...` catch all for your prop:

```tsx
const MyStack = styled(Stack, {
  variants: {
    sx: {
      '...': (style) => style
    }
  }
})
```

And then override the type signature of that MyStack to remove the style props. Now you have `<Stack sx={{ p: '$0', mx: '$10' />` or whatever tokens you want, etc.
